### ISSUE TITLE
[Performance][main-dev] Reduce direct Triton JIT host-launch latency and add a regression benchmark

### ISSUE THREAD

**Issue by rogerchen838** [triton-lang/triton-ascend#1676]
## Summary

Please reduce the steady-state host-launch latency of direct Triton JIT kernels
on `main-dev`, and add a direct C-launcher regression benchmark.

At the current `main-dev` head (`c8d95530e29ee692bf595e63bc73fe333f0dfd34`),
the generated Python extension still exposes `launch` with `METH_VARARGS` and
parses the temporary tuple through `PyArg_ParseTuple`. The hot path also performs
per-launch Python/object handling, metadata lookup, pointer extraction, launch
argument preparation, and task-queue dispatch. These fixed costs dominate small
kernels.

#1651 targets `main-dev` and contains several relevant optimizations, including
`METH_FASTCALL`, direct indexed argument parsing, interned dictionary keys,
one-time `dlsym`, zero-copy argument packing, and a reusable thread-local launch
buffer. It is still open. This issue asks for a reproducible direct-JIT
acceptance benchmark and an explicit host-enqueue regression target, whether
the final implementation is #1651 or a smaller follow-up.

This is distinct from #1675, which records measurements for
`release/3.2.2-dev`. Those numbers must not be treated as measurements of
`main-dev`.

## Baseline source observations

Baseline: `main-dev@c8d95530e29ee692bf595e63bc73fe333f0dfd34`

In `third_party/ascend/backend/driver.py`:

- the generated extension registers `launch` as `METH_VARARGS`;
- `_launch(...)` parses all launch and kernel arguments with
  `PyArg_ParseTuple`;
- tensor pointers and packed metadata are resolved on each invocation;
- the ordered task-queue path constructs and dispatches a launch callable;
- launch utility symbols are dynamically resolved and checked by the generated
  launcher.

The already-merged #1438 improves the ordered launch path, but it does not
remove all Python/C-extension fixed overhead. #1651 reports a softmax launcher
call median reduction from 65.85 us to 45.35 us, but that measurement includes
a higher-level launcher call and is not a direct raw-C-launch host-enqueue
microbenchmark. A smaller benchmark is needed to isolate the launch boundary.

## What should be measured

Measure **host enqueue time**, separately from device kernel execution time:

1. Compile one tiny direct `@triton.jit` kernel outside the timed region.
2. Obtain the resulting `CompiledKernel` and its already-built C launcher.
3. Resolve the stream and launch metadata once outside timing.
4. Warm up 1,000 launches.
5. Time 5,000 C-launcher calls without synchronizing inside the timed loop.
6. Stop the host timer, then synchronize.
7. Repeat five batches and report all samples plus their median.
8. Verify the output before measuring.

Do not call `JITFunction.run()` in the timed loop: its Python binding, cache
lookup, and specialization work are not C-launcher overhead. Do not use
Inductor or `torch.compile` for this microbenchmark.

NPU profiler kernel duration is useful as a separate device-side metric, but it
cannot replace the host timer for enqueue latency.

## Reproducer

```python
import statistics
import time

import torch
import torch_npu  # noqa: F401
import triton
import triton.language as tl


@triton.jit
def launch_probe(x, y, out, n: tl.constexpr, BLOCK: tl.constexpr):
    offsets = tl.arange(0, BLOCK)
    mask = offsets < n
    tl.store(out + offsets,
             tl.load(x + offsets, mask=mask) +
             tl.load(y + offsets, mask=mask),
             mask=mask)


def make_raw_launcher(x, y, out):
    grid = (1,)
    kernel = launch_probe.run(x, y, out, 400, BLOCK=512,
                              grid=grid, warmup=False)
    kernel._init_handles()
    driver = triton.runtime.driver.active
    stream = driver.get_current_stream(driver.get_current_device())
    metadata = kernel.launch_metadata(grid, stream, x, y, out, 400)
    args = (
        1, 1, 1, stream, kernel.function, kernel.packed_metadata, metadata,
        kernel.__class__.launch_enter_hook,
        kernel.__class__.launch_exit_hook,
        x, y, out, 400,
    )
    raw = getattr(kernel.run, "fast_launch", None)
    entry = "fast_launch" if raw is not None else "launch"
    raw = raw if raw is not None else kernel.run.launch
    return lambda: raw(*args), entry


def measure(call, warmup=1000, iters=5000, repeats=5):
    samples = []
    for _ in range(repeats):
        for _ in range(warmup):
            call()
        torch.npu.synchronize()
        start = time.perf_counter()
        for _ in range(iters):
            call()
        elapsed = time.perf_counter() - start
        torch.npu.synchronize()  # intentionally outside the timed interval
        samples.append(elapsed * 1e6 / iters)
    return statistics.median(samples), samples


x = torch.randn((4, 100), device="npu", dtype=torch.float32)
y = torch.randn_like(x)
out = torch.empty_like(x)
call, entry = make_raw_launcher(x, y, out)
call()
torch.npu.synchronize()
torch.testing.assert_close(out, x + y, rtol=1e-3, atol=1e-3)

triton_us, triton_samples = measure(call)
eager_us, eager_samples = measure(lambda: torch.add(x, y, out=out))
print(entry, triton_us, triton_samples)
print("eager", eager_us, eager_samples)
```

## Requested outcome

1. Evaluate #1651 (or an equivalent focused change) with the direct raw-C
   launcher benchmark above on `main-dev`.
2. Report separately:
   - direct C-launcher host enqueue latency;
   - the public `CompiledKernel.run` / wrapper latency;
   - NPU device kernel duration.
3. Add the direct-JIT benchmark to CI or the performance suite and retain raw
   samples, median, branch, and exact commit in its output.
4. Keep correct fallback behavior for hooks, profiling, capture, device print,
   workspace allocation, and task-queue-disabled mode.
5. Document the lifetime assumption behind any zero-copy or thread-local
   argument buffer: the dispatch must consume/copy the arguments synchronously
   before the launcher returns.

## Environment

- Hardware: 8 x Huawei Ascend 910B2; test process uses NPU 0 only
  - Product: `IT21HMDC_Bin0`; 24 AICores; 65,536 MiB HBM per card
- Driver / board software: `26.0.rc1`
- Firmware: `9.0.0.0.205`
- CANN toolkit and operator packages: `9.1.0-beta.3`
- Architecture: AArch64
- Container OS: Ubuntu 22.04.5 LTS
- Host kernel visible in the container:
  `6.6.0-132.0.0.111.oe2403sp3.aarch64`
- glibc: 2.35
- Python: 3.10.20
- PyTorch: `2.13.0+cpu`
- torch-npu: `2.13.0+git4d712d1`
- Triton-Ascend: `main-dev@c8d95530e29ee692bf595e63bc73fe333f0dfd34`

## Related work

- #1651: open host-launch optimization PR targeting `main-dev`
- #1675: direct-JIT host-launch measurements for `release/3.2.2-dev`
- #1438: JIT launch-path optimization already merged into `main-dev`

**Comment by LH-123L** [comment]
Thank you very much for your feedback. The collection of performance data is done through profiling. This PR only optimizes the launch code, so the observation point starts from driver.py->call().
This timing measurement approach is a commonly used method by users, making it more intuitive.
Regarding the CI gate, since performance testing depends on the build machine, adding a gate would affect the merging of PRs, so it will not be added for now.
As for the optimization, if there are better solutions, I welcome discussions and can be contacted directly at 00654397.

Your case test：
Before optimization:
7.025226019322872 [7.447106018662453, 7.025226019322872, 7.01014599762857, 6.960901990532875, 7.139492034912109]
6.488255970180035 [7.260424038395286, 6.239229999482632, 6.134648015722632, 6.488255970180035, 6.551842018961906]
After optimization:
5.298032006248832 [5.50916800275445, 5.334906000643969, 5.067135998979211, 5.085882032290101, 5.298032006248832]
5.007951986044645 [5.84622398018837, 5.123547976836562, 5.007951986044645, 4.970129998400807, 4.984340025112033]
